From 247fda3f39b6842915dadb55ba4f5e372bcbd444 Mon Sep 17 00:00:00 2001
From: "Kevin F. Haggerty" <haggertk@lineageos.org>
Date: Mon, 12 Nov 2018 20:23:43 -0700
Subject: [PATCH 2/2] netd: Allow devices to opt-out of the tethering active
 FTP helper

* AOSP change 4cef4147bfc1c8708c32972191690e1cd449e2e1 enabled a
  conntrack helper for active FTP connections with tethering. This
  does not play well with legacy nf_conntrack kernel implementations.
* Allow affected devices to opt-out so that tethering actually starts

Change-Id: Ic063142492eb20725e5aab1bbaebac82d74f4211
---
 server/Android.mk               | 4 ++++
 server/TetherController.cpp     | 2 ++
 server/TetherControllerTest.cpp | 4 ++++
 3 files changed, 10 insertions(+)

diff --git a/server/Android.mk b/server/Android.mk
index 3424a8c..46ddc71 100644
--- a/server/Android.mk
+++ b/server/Android.mk
@@ -72,6 +72,10 @@ ifeq ($(strip $(TARGET_NEEDS_NETD_DIRECT_CONNECT_RULE)),true)
         LOCAL_CPPFLAGS += -DNEEDS_NETD_DIRECT_CONNECT_RULE
 endif
 
+ifeq ($(strip $(TARGET_OMIT_NETD_TETHER_FTP_HELPER)),true)
+        LOCAL_CPPFLAGS += -DOMIT_NETD_TETHER_FTP_HELPER
+endif
+
 LOCAL_INIT_RC := netd.rc
 
 LOCAL_SHARED_LIBRARIES := \
diff --git a/server/TetherController.cpp b/server/TetherController.cpp
index bec8b10..2af1abd 100644
--- a/server/TetherController.cpp
+++ b/server/TetherController.cpp
@@ -638,10 +638,12 @@ int TetherController::setForwardRules(bool add, const char *intIface, const char
     }
 
     std::vector<std::string> v4 = {
+#ifndef OMIT_NETD_TETHER_FTP_HELPER
         "*raw",
         StringPrintf("%s %s -p tcp --dport 21 -i %s -j CT --helper ftp",
                      op, LOCAL_RAW_PREROUTING, intIface),
         "COMMIT",
+#endif
         "*filter",
         StringPrintf("%s %s -i %s -o %s -m state --state ESTABLISHED,RELATED -g %s",
                      op, LOCAL_FORWARD, extIface, intIface, LOCAL_TETHER_COUNTERS_CHAIN),
diff --git a/server/TetherControllerTest.cpp b/server/TetherControllerTest.cpp
index 6e77bfe..91531de 100644
--- a/server/TetherControllerTest.cpp
+++ b/server/TetherControllerTest.cpp
@@ -127,10 +127,12 @@ protected:
             "COMMIT\n", intIf);
 
         std::vector<std::string> v4Cmds = {
+#ifndef OMIT_NETD_TETHER_FTP_HELPER
             "*raw",
             StringPrintf("-A tetherctrl_raw_PREROUTING -p tcp --dport 21 -i %s -j CT --helper ftp",
                          intIf),
             "COMMIT",
+#endif
             "*filter",
             StringPrintf("-A tetherctrl_FORWARD -i %s -o %s -m state --state"
                          " ESTABLISHED,RELATED -g tetherctrl_counters", extIf, intIf),
@@ -199,10 +201,12 @@ protected:
             "COMMIT\n", intIf);
 
         std::vector<std::string> v4Cmds = {
+#ifndef OMIT_NETD_TETHER_FTP_HELPER
             "*raw",
             StringPrintf("-D tetherctrl_raw_PREROUTING -p tcp --dport 21 -i %s -j CT --helper ftp",
                          intIf),
             "COMMIT",
+#endif
             "*filter",
             StringPrintf("-D tetherctrl_FORWARD -i %s -o %s -m state --state"
                          " ESTABLISHED,RELATED -g tetherctrl_counters", extIf, intIf),
-- 
2.17.1

